### PR TITLE
Process goal conversions for device data (type, brand, model)

### PR DESCRIPTION
**Pull request by sgiehl**
rebased #10026 for 3.x
(refs #9777)

### CODE REVIEW

File: plugins/DevicesDetection/Archiver.php [hunk 1]
```diff
@@ -73,8 +73,26 @@ public function aggregateMultipleReports()
...
+            }
+
+            while ($conversionRow = $query->fetch()) {
+
+                $metrics->sumMetricsGoals($conversionRow[$labelSQL], $conversionRow);
+            }
+            $metrics->enrichMetricsWithConversions();
+        }
+
+        $report = $metrics->asDataTable()->getSerialized($this->maximumRows, null, Metrics::INDEX_NB_VISITS);
```
Comment (tsteur) [inline plugins/DevicesDetection/Archiver.php]: I would recommend to call here `$table = $metrics->asDataTable(); ... serialize ...; Common::destroy($table);`

File: plugins/DevicesDetection/Archiver.php [hunk 2]
```diff
@@ -73,8 +73,26 @@ public function aggregateMultipleReports()
...
         $this->getProcessor()->insertBlobRecord($recordName, $report);
```
Comment (tsteur) [inline plugins/DevicesDetection/Archiver.php]: and just to be safe afterwards `unset($table);unset($report);`. Otherwise we might leak memory

**Comment by tsteur** [comment]
Only needs a small change for the possible memory leaks. Otherwise feel free to merge after this change was made 👍

**Comment by sgiehl** [comment]
I've updated and rebased the branch. If the tests succeed, I'll merge it afterwards.

**Comment by mattab** [comment]
@sgiehl there is now data on demo2 :+1: 

Just noticed an issue: on demo2 at: https://demo2.piwik.org/index.php?module=CoreHome&action=index&idSite=1&period=month&date=today#?idSite=1&period=month&date=today&category=Goals_Goals&subcategory=1

 report shows conversion rates for `Unknown` device type is higher than 100%:

device type conversion rate too high

**Comment by sgiehl** [comment]
Maybe that's a more general problem. Some keywords have more than 100%, as well.

image

Had a look at the goal, but it seems it should only be triggered once a visit.

**Pull request merged**